workitem/event/event_repository.go

```diff
@@ -0,0 +1,228 @@
+	"github.com/fabric8-services/fabric8-wit/rendering"
+	"github.com/fabric8-services/fabric8-wit/workitem"
+)
+
+// APIStringTypeEvents represent the type of event
+const APIStringTypeEvents = "events"
+
+// Repository encapsulates retrieval of work item events
+type Repository interface {
+	//repository.Exister
+	List(ctx context.Context, wiID uuid.UUID) ([]Event, error)
+}
+
+// NewEventRepository creates a work item event repository based on gorm
+func NewEventRepository(db *gorm.DB) *GormEventRepository {
+	return &GormEventRepository{
+		db:               db,
+		workItemRepo:     workitem.NewWorkItemRepository(db),
+		wiRevisionRepo:   workitem.NewRevisionRepository(db),
+		workItemTypeRepo: workitem.NewWorkItemTypeRepository(db),
+		identityRepo:     account.NewIdentityRepository(db),
+	}
+}
+
+// GormEventRepository represents the Gorm model
+type GormEventRepository struct {
+	db               *gorm.DB
+	workItemRepo     *workitem.GormWorkItemRepository
+	wiRevisionRepo   *workitem.GormRevisionRepository
+	workItemTypeRepo *workitem.GormWorkItemTypeRepository
+	identityRepo     *account.GormIdentityRepository
+}
+
+// List return the events
+func (r *GormEventRepository) List(ctx context.Context, wiID uuid.UUID) ([]Event, error) {
+	revisionList, err := r.wiRevisionRepo.List(ctx, wiID)
+	if err != nil {
+		return nil, errs.Wrapf(err, "error during fetching event list")
+	}
+	if revisionList == nil {
+		return []Event{}, nil
+	}
+	wi, err := r.workItemRepo.LoadByID(ctx, wiID)
+	if err != nil {
+		return nil, errs.Wrapf(err, "error during fetching event list")
+	}
+	wiType, err := r.workItemTypeRepo.Load(ctx, wi.Type)
+	if err != nil {
+		return nil, errs.Wrapf(err, "error during fetching event list")
+	}
+
+	eventList := []Event{}
+	for k := 1; k < len(revisionList); k++ {
+		modifierID, err := r.identityRepo.Load(ctx, revisionList[k].ModifierIdentity)
+		if err != nil {
+			return nil, errs.Wrapf(err, "error during fetching event list")
+		}
+		for fieldName, field := range wiType.Fields {
+			switch fieldType := field.Type.(type) {
+			case workitem.ListType:
+				switch fieldType.ComponentType.Kind {
+				case workitem.KindLabel, workitem.KindUser:
+					var p []interface{}
+					var n []interface{}
+
+					previousValues := revisionList[k-1].WorkItemFields[fieldName]
+					newValues := revisionList[k].WorkItemFields[fieldName]
+					switch previousValues.(type) {
+					case nil:
+						p = []interface{}{}
+					case []interface{}:
+						for _, v := range previousValues.([]interface{}) {
+							p = append(p, v)
+						}
+					}
+
+					switch newValues.(type) {
+					case nil:
+						n = []interface{}{}
+					case []interface{}:
+						for _, v := range newValues.([]interface{}) {
+							n = append(n, v)
+						}
+
+					}
+
+					// Avoid duplicate entries for empty labels or assignees
+					if len(p) != 0 || len(n) != 0 {
+						fmt.Println("label", fieldName)
+						wie := Event{
+							ID:        uuid.NewV4(),
```

Comment:
The event ID we return is the same as the corresponding revision ID which will be `revisionList[k].ID`.

Refer: https://github.com/fabric8-services/fabric8-wit/pull/1959/files#diff-64c8a1eca0152efc499d40c8cac570b7R140